File: Data Structures & Algorithms/products-of-array-discluding-self/submission-3.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> l;
        int total = 1;
        for (auto& n: nums){
            l.push_back(total);
            total = total * n;
        }
        int total2 = 1;
        for (int i = nums.size()-1; i>=0; i--){
            l[i] = l[i]*total2;
            total2 = total2 * nums[i];
        }
        return l;
    }
};

File: Data Structures & Algorithms/burst-balloons/submission-3.py
class Solution:
    def maxCoins(self, nums: List[int]) -> int:
        n = len(nums)
        nums = [1] + nums + [1]
        dp = [[0] * (n + 2) for _ in range(n + 2)]
        
        for length in range(1, n + 1):
            for l in range(1, n - length + 2):
                r = l + length - 1
                for k in range(l, r + 1):
                    dp[l][r] = max(dp[l][r], dp[l][k-1] + nums[l-1] * nums[k] * nums[r+1] + dp[k+1][r])
        
        return dp[1][n]
